### ISSUE TITLE
[RFC] Refactor weight loading code

### ISSUE THREAD

**Issue by b8zhong** [sgl-project/sglang#24703]
# Weight Loader RFC

Claude assisted, I took a pass, and most of the suggestions seem reaonsable.

This RFC describes a proposal for interfaces and abstractions for the model weight loading layer.

Today, we have 165 model files in `python/sglang/srt/models/`, each carrying its own ~50–150 line `load_weights`. The largest (`deepseek_v2.py`) is 2.4k lines. And `sglang/python/sglang/srt/models/deepseek_common/deepseek_weight_loader.py` is over 700 lines long. 

Every time someone wants to fix a behavior here, they end up touching one model and missing the rest, since the same control flow is copy-pasted across the whole directory. For examples, the following PRs are for simply adding remapping per model for NVFP4 models:

https://github.com/sgl-project/sglang/pull/18224
https://github.com/sgl-project/sglang/pull/18189

But the same happens for supporting PP, etc. 

(There are more, but I cannot find them right now..)

## Motivation

The per-model `load_weights` functions repeat the same boilerplate over and over. Some quick numbers from grep today:

- 116 / 165 model files declare their own `stacked_params_mapping` (almost always the same `qkv_proj` / `gate_up_proj` tuples).
- 83 / 165 files repeat the `rotary_emb.inv_freq` / `cos_cached` / `sin_cached` skip block verbatim.
- 27 / 165 files repeat the `maybe_remap_kv_scale_name(...)` block.
- 28 / 165 files repeat the `tie_word_embeddings` + `lm_head.weight` skip.
- 352 call sites to `is_pp_missing_parameter` / `start_layer` / `end_layer` guards, all sprinkled inside the per-model loops.

So fixes to FP8 kv-scale remapping, fused module support, or PP edge cases tend to land in one or two files and quietly miss the rest. The `DeepseekV2WeightLoaderMixin` extraction is evidence that we already feel this pain — but the pattern stops at one model family.

The control flow inside each `load_weights` is fundamentally the same procedural sweep:

```text
for name, weight in weights:
    skip rotary buffers;
    skip out-of-range PP layers;
    remap kv-scale names;
    skip lm_head if tied;
    walk stacked_params_mapping → maybe load + break;
    walk expert_params_mapping → maybe load + break;
    fall-through default_weight_loader.
```

This isn't really domain logic, it's plumbing — and it would be much nicer to have it living in one place.

For reference, vLLM solved a very similar problem with `AutoWeightsLoader` (`vllm/model_executor/models/utils.py:117`). Their `llama.py:load_weights` is 5 lines. Ours is 74 lines doing essentially the same thing.

### Abstraction

Three pieces, each non-intrusive:

1. **Walker** (`AutoWeightsLoader`). Given a root `nn.Module` and an iterable of `(name, tensor)`, it walks the module tree. For each child it finds, if that child has its own `load_weights`, defer to it; otherwise recurse; at the leaf, look up `param.weight_loader` (already the convention used by sglang's linear / fused MoE layers) and call it. This is where rotary skip, PP skip, kv-scale remap, tie-word-embeddings skip, and "not found" diagnostics all live exactly once.

2. **Name remapper** (`WeightsMapper`). A small declarative dataclass: `{regex|substr|prefix|suffix → new_name | None}`. Replaces the dozens of `name = name.replace("foo", "bar")` lines, plus the "set name to None to drop it" idiom. Composable via `|` so quant remaps and model remaps stack cleanly.

3. **Quirks registry**, parallel to the `OverrideRegistry` in the server-args proposal. Per-architecture hooks for things like DeepSeek's `fused_qkv_a_proj_with_mqa` (currently injected into `packed_modules_mapping` inside `__init__`) or w4afp8's expert-input-scale mapping. Today these are `if quant_config.get_name() == "w4afp8": ...` branches inside the loader; they should be registered, not branched, in the model loading code.

### Interface

After the walker is in place, a typical model's `load_weights` would look like:

```python
class LlamaForCausalLM(nn.Module):
    def load_weights(self, weights):
        loader = AutoWeightsLoader(
            self,
            skip_prefixes=["lm_head."] if self.config.tie_word_embeddings else None,
        )
        return loader.load_weights(weights)
```

That covers the 80% case. Models with real special cases override at the *module* level instead of the top-level model:

```python
class DeepseekV2MoE(nn.Module):
    def load_weights(self, weights):
        # MoE-specific expert sharding — only this submodule, not the whole model
        ...
```

The walker sees `DeepseekV2MoE.load_weights` exists and defers to it; the rest of the model still rides the auto path. This is where most of the win comes from: today, one quirk in one submodule forces you to write the entire top-level loader by hand.

```python
@WeightQuirkRegistry.register(["DeepseekV3ForCausalLM"])
def deepseek_quirks(model, mapper: WeightsMapper) -> WeightsMapper:
    if model.fuse_qkv_a_proj:
        mapper |= WeightsMapper(orig_to_new_substr={
            "q_a_proj": "fused_qkv_a_proj_with_mqa",
            "kv_a_proj_with_mqa": "fused_qkv_a_proj_with_mqa",
        })
    return mapper
```

Subsystem interface: nothing else in the codebase needs to know the walker exists. `model_loader/loader.py` keeps calling `model.load_weights(...)` exactly as today, and the change is local to `srt/models/*`.

## Breaking Changes

Since this is not exactly **public** API, we can maybe keep on making further changes for readability/usability/ease of maintenance. It will only affect those with out of tree models (requires them changing the mdoel code).

- Delete `stacked_params_mapping` and `packed_modules_mapping` from every model class. They move to the layer that owns the fusion. Out-of-tree models that copy-pasted these will need to drop them.
- Standardize the `load_weights` return type to `set[str]` (loaded param names). Today some return `None`, some return nothing. This unblocks "expected but missing" diagnostics in the walker.
- Delete `do_load_weights` / `DeepseekV2WeightLoaderMixin` — the walker subsumes it. Submodule-level `load_weights` overrides replace it cleanly.
- Delete `get_module_name_from_weight_name`, `get_num_params`, `get_weights_by_name` from individual model classes (used mostly by tests and weight-update RPC). These are mechanical and belong on the walker / a single helper, not duplicated per model.
- Drop the per-loop `is_pp_missing_parameter` / `start_layer` / `end_layer` checks. Make `PPMissingLayer` short-circuit traversal in the walker so model code never has to think about PP boundaries again. (~352 call sites go away.)
- Lift the `ThreadPoolExecutor` + `should_async_load` path out of the DeepSeek mixin and make it the default in the walker, so every model gets it for free. Behavior change: load order becomes non-deterministic, but `weight_loader` callers should already not depend on order.
- Standardize on `Iterable[tuple[str, torch.Tensor]]` (drop the legacy `Tuple` typing, drop the variant signatures like `is_nextn=False` kwarg — the NextN distinction can be handled by a quirk that swaps in a different mapper).
- `bitsandbytes_stacked_params_mapping` becomes the same registry mechanism, not a separate per-model attribute.

These are aggressive but each one removes a category of duplication, not just lines.

## Open questions

1. Migration order. first, do the easy and less used models, we can do Deepseek files at the very end.

**Comment by JD-ETH** [comment]
thanks for the proposal! I also feel strongly that this is needed. In my P2P [weight transfer](https://www.lmsys.org/blog/2026-04-29-p2p-update/) implementation, I have made very similar experience already, but didn't manage to push my changes to main: https://github.com/sgl-project/sglang/pull/17326 

some additional points to yours, that claude has missed: 
- weight mapping logic is not one to one, but `one-to-shard` --- Knowing what exactly changed can be very informative when loading data 
- beyond the standard stacked_parameter_mapping, there is very error-prone "_in-function-cache_" like [this](https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/models/deepseek_common/deepseek_weight_loader.py#L293) , which does not persistent between load_weight calls and can lead to broken weight load. 
- `post_weight_load` hook typically handles tensors that are not part of `model_parameters` dictionary and are hardware specific, and should be handeled independently completely, like the `w_kc`, `w_vc` in `MLA`. 


I'd be willing to help on this ---- model loading has been a huge pain point, and once we properly unify the surfaces we can implement more generic "streamed" data loader functionalities that does not all open the whole hugging face weight file for every rank on the host machine.

**Comment by b8zhong** [comment]
Hi @JD-ETH , thanks. I agree with:
- the first point

> there is very error-prone "in-function-cache" like [this](https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/models/deepseek_common/deepseek_weight_loader.py#L293) , which does not persistent between load_weight calls and can lead to broken weight load.
I think these should declared by the layer itself, to avoid these kind of bugs.

And for the third one, we can have it as a hook, to avoid polluting weight loading code.

**Comment by b8zhong** [comment]
@JD-ETH Can you message me on Slack to discuss more? I'm @Brayden Zhong

**Comment by JD-ETH** [comment]
@b8zhong  did you see my ping on slack?

**Comment by b8zhong** [comment]
Discuss with @JD-ETH :


> I wonder however whether this can be dropped completely ---- is there a strict dependency for them to all be available first before a shard can be directly updated for fused_qkv_a_proj_with_mqa ? can't we just update the corresponding shard directly by updating the WeightLoader there instead?
I have a strong preference for the load_weights signature to not only including the set of new names, but also which corresponding shard it was updated (tensor and expert shard): layer0.q_proj -> layer0.qkv_proj[q]
Is there a way to provide a native assertion helper, so that caller knows whether all expected weights had been updated? For example, the caller can then do 
```python
updated = {}
for i in range(update_steps): 
   updated += model.load_weights(weights_part_i)
assert set(updated) == model.all_parameters(), "Weight update incomplete, missing {model.all_parameters() - updated}"
```
> where all_parameters() is a combination of named_parameters + additional parameters (like fused_qkv_a_proj_with_mqa) minus stuff like rotary_pos_emb. Note that this would require each shard to have different name to be correct.
This  gives the load_weights output signature a clear use case and a clear improvement over existing weight load logic.

We will do the above.
> "quirk" is hard to understand for me, can we name it for what it does WeightNameRemapRegistry ?
We're renaming it to RemapRegistry

**Comment by JD-ETH** [comment]
thanks! Yes, I plan to build an initial skeleton for the new weight loading common abstractions, and ask for reviews, before extending the interface to all existing models.

**Comment by Fridge003** [comment]
@b8zhong Nice design, I agree with the overall idea. But can you please provide some details, including:
1. The members variables/methods  of AutoWeightsLoader, WeightsMapper and OverrideRegistry. If it's already in the codebase (like `AutoWeightsLoader`), do we add or change anything. 
2. How to ensure backward compatibility, for example, use an environ `SGLANG_ENABLE_WEIGHT_LOADER_V2` and protect all the refactored codes under this path. Weight loader is sensitive to modeling accuracy so we need to do this
3. How are you going to test these newly added abstractions. I would prefer some unit tests that don't cost much time to run.
4. Maybe open a draft PR that implements your ideas and run things correctly. It can be vibed with gpt/claude, and we can break it into smaller parts on its basis.

**Comment by Fridge003** [comment]
Also, there are some advanced loading techniques, like multithread loading or nfs prefetching. We need to inspect the compatibility under the new design

**Comment by JD-ETH** [comment]
good point, let me build a minimal example first. thanks @Fridge003 !

**Comment by ch-wan** [comment]
FYI, this PR includes a checksum function that would be useful for verifying the correctness of weight loading. https://github.com/sgl-project/sglang/pull/24256

**Comment by b8zhong** [comment]
1. We technically already have AutoWeightsLoader, which is from the HF transformers model definition. Although, currently it's a vLLM copy, and it's only for Transformers modelling backends. We should consolidate these two if possible (currently deciding on the member vars/funcs)

2. For backward compatibility, should we use an env var? Since, the weight loading code for each model will only change once we refactor it's modelling file. For example for the most important models like Dpsk V2, we can be the most careful of it's usage and test with checksum

3. We can use unit tests, for each class depending on it's purpose, and cover different model files with various quantization methods, etc.
4. For a draft PR, https://github.com/sgl-project/sglang/compare/main...brayden/weight-loader-refactor-part-1 needs to be updated with our more up to date plan

**Comment by JD-ETH** [comment]
https://github.com/sgl-project/sglang/blob/2ba329218d1eee9a4fa05d053a3f3d666472384f/python/sglang/srt/model_loader/design.md

**Comment by JD-ETH** [comment]
Implementation/migration tracker for this RFC: #31051 (PR1–PR9 plan). Demo PR: https://github.com/sgl-project/sglang/pull/28671
